src/main/scala/rocket/dcache.scala

```diff
@@ -495,60 +497,83 @@ class DCache(implicit p: Parameters) extends L1HellaCacheModule()(p) {
+    when (io.dmem.req.fire()) { state := s_wait }
+
+    val acq = Reg(tl_in.a.bits)
+    when (io.dmem.resp.valid) { acq.data := io.dmem.resp.bits.data }
+    when (tl_in.a.fire()) { acq := tl_in.a.bits }
+
+    val isWrite = edge.hasData(acq)
+    val isRead = !isWrite
+
+    def formCacheReq(acq: TLBundleA) = {
+      val req = Wire(new HellaCacheReq)
+      // treat all loads as full words, so bytes appear in correct lane
+      req.typ := Mux(isRead, log2Ceil(beatBytes), acq.size)
+      req.cmd := Mux(isRead, M_XRD, M_XWR)
+      req.addr := Mux(isRead, ~(~acq.address | (beatBytes-1)), acq.address)
+      req.tag := UInt(0)
+      req
+    }
 
-  val ready = state === s_ready || io.tl.grant.fire()
-  io.dmem.req.valid := (io.tl.acquire.valid && ready) || state === s_replay
-  io.tl.acquire.ready := io.dmem.req.ready && ready
-  io.dmem.req.bits := formCacheReq(Mux(state === s_replay, acq, io.tl.acquire.bits))
-  // this blows.  the TL data is already in the correct byte lane, but the D$
-  // expects right-justified store data, so that it can steer the bytes.
-  io.dmem.s1_data := new LoadGen(acq.op_size(), Bool(false), acq.addr_byte(), acq.data, Bool(false), acq.tlDataBytes).data
-  io.dmem.s1_kill := false
-  io.dmem.invalidate_lr := false
-
-  // place AMO data in correct word lane
-  val minAMOBytes = 4
-  val grantData = Mux(io.dmem.resp.valid, io.dmem.resp.bits.data, acq.data)
-  val alignedGrantData = Mux(acq.op_size() <= log2Ceil(minAMOBytes), Fill(coreDataBytes/minAMOBytes, grantData(8*minAMOBytes-1, 0)), grantData)
-
-  io.tl.grant.valid := io.dmem.resp.valid || state === s_grant
-  io.tl.grant.bits := Grant(
-    is_builtin_type = Bool(true),
-    g_type = acq.getBuiltInGrantType(),
-    client_xact_id = acq.client_xact_id,
-    manager_xact_id = UInt(0),
-    addr_beat = acq.addr_beat,
-    data = alignedGrantData)
+    val ready = state === s_ready || tl_in.d.fire()
+    io.dmem.req.valid := (tl_in.a.valid && ready) || state === s_replay
+    tl_in.a.ready := io.dmem.req.ready && ready
+    io.dmem.req.bits := formCacheReq(Mux(state === s_replay, acq, tl_in.a.bits))
+    // this blows.  the TL data is already in the correct byte lane, but the D$
+    // expects right-justified store data, so that it can steer the bytes.
+    io.dmem.s1_data := new LoadGen(acq.size, Bool(false), acq.address(log2Ceil(beatBytes)-1,0), acq.data, Bool(false), beatBytes).data
+    io.dmem.s1_kill := false
+    io.dmem.invalidate_lr := false
+
+    // place AMO data in correct word lane
+    val minAMOBytes = 4
+    val grantData = Mux(io.dmem.resp.valid, io.dmem.resp.bits.data, acq.data)
+    val alignedGrantData = Mux(acq.size <= log2Ceil(minAMOBytes), Fill(coreDataBytes/minAMOBytes, grantData(8*minAMOBytes-1, 0)), grantData)
+
+    tl_in.d.valid := io.dmem.resp.valid || state === s_grant
+    tl_in.d.bits := Mux(isRead,
+      edge.AccessAck(acq, UInt(0), alignedGrantData),
+      edge.AccessAck(acq, UInt(0)))
```

Comment:
alignedGrantData should be selected for AMOs as well as reads.  As is, it looks like isRead will be false in that case, so AMO data won't be returned.

Can we fix the bug by deleting the Mux and always return edge.AccessAck(acq, UInt(0), alignedGrantData)?